WordPress/src/main/java/org/wordpress/android/ui/accounts/login/nav/LoginNavController.java

```diff
@@ -0,0 +1,150 @@
+            return new InputSiteAddressHandler();
+        }
+
+        throw new RuntimeException("Unsupported login state " + loginNav.getSimpleName());
+    }
+
+    private Stack<Class<? extends LoginNav>> mLoginNavStack = new Stack<>();
+
+    private boolean isInState(Class<? extends LoginNav> loginNav) {
+        return !mLoginNavStack.empty() && loginNav.equals(mLoginNavStack.peek());
+    }
+
+    public void ensureState(Class<? extends LoginNav> loginNav) {
+        if (!isInState(loginNav)) {
+            throw new RuntimeException("Not in state " + loginNav.getSimpleName());
+        }
+    }
+
+    private void gotoState(Class<? extends LoginNav> loginNav) {
+        if (!isInState(loginNav)) {
+            mLoginNavStack.push(loginNav);
+        }
+    }
+
+    public void goBack() {
+        if (mLoginNavStack.isEmpty()) {
+            throw new RuntimeException("Navigation stack is empty! Can't go back.");
+        }
+
+        mLoginNavStack.pop();
+    }
+
+    public boolean isNavStackEmpty() {
+        return mLoginNavStack.isEmpty();
+    }
+
+    // available for testing purposes. Don't use otherwise
+    public void force(Class<? extends LoginNav> loginNav) {
+        mLoginNavStack.push(loginNav);
+    }
+
+    private class PrologueHandler implements LoginNav.Prologue {
+        @Override
+        public void doStartLogin() {
+            ensureState(LoginNav.Prologue.class);
+            gotoState(LoginNav.InputEmail.class);
+
+            mLoginNavHandler.showEmailLoginScreen();
+        }
+
+        @Override
+        public void doStartSignup() {
+            ensureState(LoginNav.Prologue.class);
+            gotoState(LoginNav.Prologue.class);
+
+            mLoginNavHandler.toast("Signup is not implemented yet");
+        }
+    }
+
+    private class InputEmailHandler implements LoginNav.InputEmail {
+        @Override
+        public void gotEmail(String email) {
+            ensureState(LoginNav.InputEmail.class);
+            gotoState(LoginNav.InputEmail.class);
+
+            mLoginNavHandler.toast("Input email is not implemented yet. Input email: " + email);
+        }
+
+        @Override
+        public void loginViaUsernamePassword() {
+            ensureState(LoginNav.InputEmail.class);
+            gotoState(LoginNav.InputEmail.class);
+
+            mLoginNavHandler.toast("Fall back to username/password is not implemented yet.");
+        }
+
+        @Override
+        public void help() {
+            ensureState(LoginNav.InputEmail.class);
+            gotoState(LoginNav.InputEmail.class);
+
+            mLoginNavHandler.toast("Help is not implemented yet.");
+        }
+    }
+
+    private class InputSiteAddressHandler implements LoginNav.InputSiteAddress {
+        @Override
+        public void gotSiteAddress(String siteAddress) {
+            ensureState(LoginNav.InputSiteAddress.class);
+            gotoState(LoginNav.InputSiteAddress.class);
```

Comment:
I meant to comment about this in my previous review and forgot.

The `ensureState` has the following check:
```
if (!isInState(loginNav)) {
    throw new RuntimeException("Not in state " + loginNav.getSimpleName());
}
```
And in `gotoState` we have this:
```
if (!isInState(loginNav)) {
    mLoginNavStack.push(loginNav);
}
```
So, that means we'll either throw an exception or the `gotoState` won't do anything since we pass `LoginNav.InputSiteAddress.class` to both methods, right?

I think this is the case in many methods except for `doStartLogin` where we pass different states to `ensureState` & `gotoState`.

If I understand it correctly, the `ensureState` checks the current "page" and `gotoPage` will change the "page" (I know they are not actually pages), so I'd imagine we'd never want to pass the same state. Am I mistaken about that?